# HW 3: Introduction and Setup
* Copy the contents of the file "transforms_hw03.py" to the bottom of your "transforms.py" file.
* Complete the function definitions for "se3" and "inv" 
* Run this notebook and check the outputs to make sure everything matches. If your code is correct, the output should exactly match the homogenous transformation matrices shown below. 
* Make sure to review each cell and think about what the operation means. Does it make sense? If not, ask about it on Piazza or in office hours. 

In [1]:
import transforms as tr
import numpy as np
from visualization import VizScene 
import time

## Homogeneous matrix made of only rotation about z-axis by amount $\frac{\pi}{4}$

Rotation by amount theta around the z-axis should give the following:
$$
\left[\begin{matrix}0.707 & -0.707 & 0 & 0\\ 0.707 & 0.707 & 0 & 0\\0 & 0 & 1 & 0\\0 & 0 & 0 & 1\end{matrix}\right]
$$

In [2]:
T = tr.se3(tr.rotz(np.pi/4))
print(T)


[[0.707106781186548 -0.707106781186547 0 0]
 [0.707106781186547 0.707106781186548 0 0]
 [0 0 1 0]
 [0 0 0 1]]


## Translation only along the x-axis by an amount $0.5$ should give the following:
$$
\left[\begin{matrix}1 & 0 & 0 & 0.5 \\0 & 1 & 0 & 0\\0 & 0 & 1 & 0\\0 & 0 & 0 & 1\end{matrix}\right]
$$

In [3]:
T = tr.se3(p=[0.5, 0, 0])
print(T)

[[1.  0.  0.  0.5]
 [0.  1.  0.  0. ]
 [0.  0.  1.  0. ]
 [0.  0.  0.  1. ]]


# Checking the "inv" function

If we calculate a homogenous transform with a rotation of 45 degrees about the x-axis and a general translation ($[0.5, 0.25, 0.3]^T$) we get the following:
$$
\left[\begin{matrix}1 & 0 & 0 & 0.5\\0 & 0.707106781186548 & -0.707106781186547 & 0.25\\0 & 0.707106781186547 & 0.707106781186548 & 0.3\\0 & 0 & 0 & 1\end{matrix}\right]
$$

Then, we can multiply the original transform by its inverse to check out inverse function, we should get the following:
$$
\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1.0 & 0 & 0\\0 & 0 & 1.0 & 0\\0 & 0 & 0 & 1\end{matrix}\right]
$$

In [4]:
T = tr.se3(tr.rotx(np.pi/4), [0.5, 0.25, 0.3])
print(T)

# now we can check if we implemented "inv" correctly:
check = tr.inv(T) @ T
print("\n\n inv(T) @ T should give identity matrix:")
print(check)


[[1 0 0 0.5]
 [0 0.707106781186548 -0.707106781186547 0.25]
 [0 0.707106781186547 0.707106781186548 0.3]
 [0 0 0 1]]


 inv(T) @ T should give identity matrix:
[[1.0 0 0 0.0]
 [0 1.00000000000000 0 0]
 [0 0 1.00000000000000 0]
 [0 0 0 1.0]]


# Transform a point in one frame to another frame


## DH parameters combined for a single joint:

Remember that if we combine a rotation in z, translation in z, then translation in x, and rotation in x, we should get the same result as the book for following the DH convention to move from one frame (or joint) to another as follows: 

$$
\left[\begin{matrix}\cos{\left(\theta \right)} & - \sin{\left(\theta \right)} \cos{\left(\alpha \right)} & \sin{\left(\alpha \right)} \sin{\left(\theta \right)} & a \cos{\left(\theta \right)}\\\sin{\left(\theta \right)} & \cos{\left(\alpha \right)} \cos{\left(\theta \right)} & - \sin{\left(\alpha \right)} \cos{\left(\theta \right)} & a \sin{\left(\theta \right)}\\0 & \sin{\left(\alpha \right)} & \cos{\left(\alpha \right)} & d\\0 & 0 & 0 & 1\end{matrix}\right]
$$

In future homework, we'll implement representations of robot arms that include this transformation. But for the test values of DH parameters below, we can perform this sequence of operations manually.  

Assuming the following DH parameters for two joints:

$\theta_1 = \frac{\pi}{8}$, $d_1 = 0$, $a_1 = 0.3$, $\alpha_1 = \frac{\pi}{2}$

$\theta_2 = \frac{\pi}{4}$, $d_2 = 0$, $a_2 = 0.3$, $\alpha_2 = 0$


The resulting homogeneous transform describing the tip (or frame 2) relate to frame 0 would be:

$$
\left[\begin{matrix}0.653281482438188 & -0.653281482438188 & 0.38268343236509 & 0.473148304484842\\0.270598050073099 & -0.270598050073099 & -0.923879532511287 & 0.195984444731456\\0.707106781186548 & 0.707106781186548 & 0 & 0.212132034355964\\0 & 0 & 0 & 1.0\end{matrix}\right]
$$

In [1]:
import transforms as tr
import numpy as np
from visualization import VizScene 
import time

# start by substituting the actual values for R and p and making a new SE3 object
# that describes the transformation from frame 0 to frame 1

# TODO - fill this out 
# find the transformation from frame 0 to 1
T1_in_0 = np.array([[.924,0,.383,.277],[.383,0,-.924,.115],[0,1,0,0],[0,0,0,1]])

# do the same thing for frame 1 to frame 2
T2_in_1 = np.array([[.707,-.707,0,.212],[.707,.707,0,.212],[0,0,1,0],[0,0,0,1]])

# now we can combine the two to get a transformation that describes frame 2 
# relative to frame 0
T2_in_0 = T1_in_0 @ T2_in_1


# printing the result
print(T2_in_0)

# use the "add_frame" function to plot both frames (for joint 1 and joint 2) relative 
# to a base or ground frame. 

# TODO - put your visualization code here. 

Tw_to_frame1 = np.array([[1, 0,  0,  1],
                         [0, 0, -1,  0],
                         [0, 1,  0,  0],
                         [0, 0,  0,  1]])
                         
viz = VizScene()
viz.add_frame(np.eye(4), label='world')
viz.add_frame(T2_in_1, label = "T2_in_1")
viz.add_frame(T1_in_0, label = "T1_in_0")

time_to_run = 10
refresh_rate = 60
for i in range(refresh_rate * time_to_run):
    t = time.time()
    Tw_to_frame1 = np.array([[np.cos(np.pi/2*t), -np.sin(np.pi/2*t), 0, 1],
                             [np.sin(np.pi/2*t), np.cos(np.pi/2*t), 0, 0],
                             [0, 0, 1, 0],
                             [0, 0, 0, 0]])
    viz.update(As=[np.eye(4), Tw_to_frame1])
    time.sleep(1.0/refresh_rate)

viz.close_viz()


[[ 0.653268 -0.653268  0.383     0.472888]
 [ 0.270781 -0.270781 -0.924     0.196196]
 [ 0.707     0.707     0.        0.212   ]
 [ 0.        0.        0.        1.      ]]


IndexError: list index out of range

: 

In [6]:
""

''